src/lib.rs

```diff
@@ -99,17 +102,61 @@ impl Diagnostic {
     }
 }
 
-pub type Sources<'a> = [&'a dyn SourceContainer];
+pub type Sources<'a> = [&'a mut dyn SourceContainer];
 
 /// SourceContainers offer source-code to be compiled via the load_source function.
 /// Furthermore it offers a location-String used when reporting diagnostics.
 pub trait SourceContainer {
     /// loads and returns the SourceEntry that contains the SourceCode and the path it was loaded from
-    fn load_source(&self) -> Result<SourceCode, String>;
+    fn load_source(&mut self, encoding: Option<&'static Encoding>) -> Result<SourceCode, String>;
```

Comment:
Can you explain why this suddenly gets mut when we introduce encoding?

Reply:
read() expects a mutable reader, so if we want to implement for a "Read" trait, we need it to be mutable.
I'll try to look into another strategy here.

Reply:
That said, the only reason I support Read here is for the tests. So I could probably move this into integration tests that load files instead.